File: Python/DinnerPlateStacks.py
import heapq
from typing import List


class DinnerPlates:
    def __init__(self, capacity: int):
        self.cap = capacity
        self.stacks: List[List[int]] = []
        self.avail: List[int] = []  # min-heap of indices with room (may hold stale entries)

    def push(self, val: int) -> None:
        # Drop stale indices: out of range or already full.
        while self.avail and (self.avail[0] >= len(self.stacks)
                              or len(self.stacks[self.avail[0]]) >= self.cap):
            heapq.heappop(self.avail)
        if not self.avail:
            heapq.heappush(self.avail, len(self.stacks))
            self.stacks.append([])
        i = self.avail[0]
        self.stacks[i].append(val)
        if len(self.stacks[i]) >= self.cap:
            heapq.heappop(self.avail)

    def pop(self) -> int:
        while self.stacks and not self.stacks[-1]:
            self.stacks.pop()
        if not self.stacks:
            return -1
        return self.popAtStack(len(self.stacks) - 1)

    def popAtStack(self, index: int) -> int:
        if index >= len(self.stacks) or not self.stacks[index]:
            return -1
        heapq.heappush(self.avail, index)  # this stack now has room
        return self.stacks[index].pop()


def run(ops, args):
    out = []
    d = None
    for op, a in zip(ops, args):
        if op == "DinnerPlates":
            d = DinnerPlates(a[0])
            out.append("null")
        elif op == "push":
            d.push(a[0])
            out.append("null")
        elif op == "pop":
            out.append(str(d.pop()))
        elif op == "popAtStack":
            out.append(str(d.popAtStack(a[0])))
    return "[" + ", ".join(out) + "]"


if __name__ == '__main__':
    ops = ["DinnerPlates", "push", "push", "push", "push", "push", "popAtStack",
           "push", "push", "popAtStack", "popAtStack", "pop", "pop", "pop", "pop", "pop"]
    args = [[2], [1], [2], [3], [4], [5], [0], [20], [21], [0], [2], [], [], [], [], []]
    print(run(ops, args))
    # [null, null, null, null, null, null, 2, null, null, 20, 21, 5, 4, 3, 1, -1]

File: CPP/DinnerPlateStacks.cpp
#include <iostream>
#include <queue>
#include <string>
#include <vector>

class DinnerPlates {
public:
    explicit DinnerPlates(int capacity) : cap(capacity) {}

    void push(int val) {
        // Drop stale indices: out of range or already full.
        while (!avail.empty() &&
               (avail.top() >= (int)stacks.size() || (int)stacks[avail.top()].size() >= cap)) {
            avail.pop();
        }
        if (avail.empty()) {
            avail.push((int)stacks.size());
            stacks.emplace_back();
        }
        int i = avail.top();
        stacks[i].push_back(val);
        if ((int)stacks[i].size() >= cap) avail.pop();
    }

    int pop() {
        while (!stacks.empty() && stacks.back().empty()) stacks.pop_back();
        if (stacks.empty()) return -1;
        return popAtStack((int)stacks.size() - 1);
    }

    int popAtStack(int index) {
        if (index >= (int)stacks.size() || stacks[index].empty()) return -1;
        avail.push(index);
        int v = stacks[index].back();
        stacks[index].pop_back();
        return v;
    }

private:
    int cap;
    std::vector<std::vector<int>> stacks;
    std::priority_queue<int, std::vector<int>, std::greater<int>> avail;  // min-heap of indices
};

int main() {
    std::vector<std::string> ops = {"DinnerPlates", "push", "push", "push", "push", "push",
        "popAtStack", "push", "push", "popAtStack", "popAtStack", "pop", "pop", "pop", "pop", "pop"};
    std::vector<int> args = {2, 1, 2, 3, 4, 5, 0, 20, 21, 0, 2, 0, 0, 0, 0, 0};

    DinnerPlates* d = nullptr;
    std::string out = "[";
    for (size_t i = 0; i < ops.size(); ++i) {
        if (i) out += ", ";
        const std::string& op = ops[i];
        if (op == "DinnerPlates") {
            delete d;
            d = new DinnerPlates(args[i]);
            out += "null";
        } else if (op == "push") {
            d->push(args[i]);
            out += "null";
        } else if (op == "pop") {
            out += std::to_string(d->pop());
        } else {  // popAtStack
            out += std::to_string(d->popAtStack(args[i]));
        }
    }
    out += "]";
    std::cout << out << "\n";
    delete d;
    return 0;
}
